### PR TITLE
Add RELEASE_PROCEDURE.md

### PR DESCRIPTION
**Pull request by oschaaf**
Prerequisite:

- [x] Land #261 to update Envoy with structured versioning support
- [x] Land #226 as that brings `version_info.h`
- [x] Merge master here after doing so

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

File: RELEASE_PROCEDURE.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    Ideally these have been tracked on the go but it does not hurt to ask around and double
+    check. Running the python integration tests of the previous release with the latest binaries could serve as a double check for this.
+  - Ensure the release notes are complete. Ideally these have been tracked on the go.
+- Perform some thorough testing to double down on stability.
+
+```bash
+bazel test --cache_test_results=no --test_env=ENVOY_IP_TEST_VERSIONS=all --runs_per_test=1000 --jobs 50 -c dbg --local_resources 20000,20,0.25 //test:*
+```
+
+- Draft a [GitHub tagged release](https://github.com/envoyproxy/nighthawk/releases/new) using the [current version number on master]((include/nighthawk/common/version.h)). Earlier releases are tagged like `v0.1` so currently we use that as a naming convention.
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: Curious - didn't we want to use semantic versioning (MAJOR.MINOR.PATCH)? I see that Envoy does use it.

(https://semver.org/spec/v2.0.0.html)

Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Oh yeah good catch, starting the upcoming 0.3 that is the plan I guess, I’ll amend

Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Amended

**Comment by oschaaf** [comment]
performed a full iteration on the proposed release procedure in https://github.com/envoyproxy/nighthawk/pull/232/commits/2aee23e1f8ff2aa9f7b5749858beba3459334e2e, this is ready for another look

File: RELEASE_PROCEDURE.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    check. Running the python integration tests of the previous release with the latest binaries could serve as a double check for this.
+  - Ensure the release notes are complete. Ideally these have been tracked on the go.
+- Based on our [semantic versioning](https://semver.org/spec/v2.0.0.html) strategy, our release numbers look like `MAJOR.MINOR.PATCH`. Determine the release type to come up with what the next version should look like:
+  - Breaking changes increment `MAJOR`, and reset `MINOR.PATCH` to `0.0`
+  - Added functionality which is backwards compatible increments `MINOR`, and resets `PATCH` to `0`
+  - Bug fixes increment `PATCH`
+
+## Release steps
+
+1. Speculatively bump the version in [version.h](include/nighthawk/common/version.h) to the version you determined earlier. This may result in version gaps if a release attempt fails, but avoids having to freeze merges to master and/or having to work with release branches. In short it helps keeping the release procedure lean and mean and eliminates the need for blocking others while this procedure is in-flight.
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: version.h is coming in a separate PR that may be delayed. Do we want to pull it in here?

File: RELEASE_PROCEDURE.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+2. Close the voting issue, if any. If needed share context in the issue so everyone has context.
+3. Subsequently re-spin the full procedure from the top after the blocking issue is resolved
+
+[1] Consider running the unit tests and integration tests repeatedly and concurrently for while.
+It's worth noting that some of the integration tests have not been designed for this, so that part
+is not tried and proven yet. (some of the integration tests are sensitive to timing, and may get
+unstable when slowed down due to concurrent runs). A sample command to do this:
+
+```bash
+bazel test --cache_test_results=no --test_env=ENVOY_IP_TEST_VERSIONS=all --runs_per_test=1000 --jobs 50 -c dbg --local_resources 20000,20,0.25 //test:*
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: (optional) Consider breaking this example into lines for easier reading experience (without the need to scroll).

```
bazel test \
  --cache_test_results=no \
  --test_env=ENVOY_IP_TEST_VERSIONS=all \
  --runs_per_test=1000 \
  --jobs 50 \
  -c dbg \
  --local_resources 20000,20,0.25 \
  //test:*
```

File: RELEASE_PROCEDURE.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  - Breaking changes increment `MAJOR`, and reset `MINOR.PATCH` to `0.0`
+  - Added functionality which is backwards compatible increments `MINOR`, and resets `PATCH` to `0`
+  - Bug fixes increment `PATCH`
+
+## Release steps
+
+1. Speculatively bump the version in [version.h](include/nighthawk/common/version.h) to the version you determined earlier. This may result in version gaps if a release attempt fails, but avoids having to freeze merges to master and/or having to work with release branches. In short it helps keeping the release procedure lean and mean and eliminates the need for blocking others while this procedure is in-flight.
+2. Draft a [GitHub tagged release](https://github.com/envoyproxy/nighthawk/releases/new). Earlier releases are tagged like `v0.1`, but as of `0.3.0`we are using [semantic versioning](https://semver.org/spec/v2.0.0.html)
+3. Perform thorough testing of the targeted revision to double down on stability [1]
+4. Create an optimized build for comparing with the previous release. Changes in performance
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: As someone who came onboard recently, this doesn't provide enough detail to know how we do the comparing. Anything else we can add here to help?

File: RELEASE_PROCEDURE.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  and/or measurement accuracy may need to be considered breaking. If in doubt, discuss!
+5. If things look good, [File an issue](https://github.com/envoyproxy/nighthawk/issues/new?title=[VOTE]+Release+v0.x&body=Release%20X%20is%20ready%20for%20review.%20Please%20take%20a%20look%20and%20vote!) to raise a vote on publishing the drafted release, and point contributors to that.
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: (nit) Shall we update the link to include a semantic version number (e.g. v0.X.X).

File: RELEASE_PROCEDURE.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+6. Allow sufficient time to transpire for others to provide feedback, before moving on to the next step. If nobody raises blocking issues, it is time to go ahead and publish the drafted release.
+7. |o/ Announce the new release on Slack, Twitter
```
Comment (mum4k) [inline RELEASE_PROCEDURE.md]: (optional) Should we include links to the Slack workspace and Twitter?

File: RELEASE_PROCEDURE.md (hunk 2 above)
Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Good point, added a pre-requisite to land this in the [PR description](https://github.com/envoyproxy/nighthawk/pull/232#issue-353280027)

Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Also tagged this is blocked for that reason

File: RELEASE_PROCEDURE.md (hunk 3 above)
Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Done in 696410260a3dcfab4d9b48823fcfcff6ddc37b00

File: RELEASE_PROCEDURE.md (hunk 4 above)
Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Punted to TODO referencing #245. For now I'd manually do some simple low and high qps testing, ensuring throughput and reported numbers are ballpark the same

File: RELEASE_PROCEDURE.md (hunk 5 above)
Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Done in 696410260a3dcfab4d9b48823fcfcff6ddc37b00

File: RELEASE_PROCEDURE.md (hunk 6 above)
Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Done in 696410260a3dcfab4d9b48823fcfcff6ddc37b00, except Twitter, as there is no official channel.

Comment (mum4k) [reply RELEASE_PROCEDURE.md]: (optional) Would it help if we provide a link to some previous tweets?

Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: That is hard, speaking for myself I didn’t do this for the previous releases yet, but maybe we can copycat Envoy, assuming it does this.. A project account would be nicest IMHO? Not sure how to approach materializing that though.

Comment (mum4k) [reply RELEASE_PROCEDURE.md]: Shall we drop Twitter from this document for now and add an issue to track this down? Not sure how much value it provides without the instructions...

Comment (oschaaf) [reply RELEASE_PROCEDURE.md]: Yeah, makes sense, done and done.

**Pull request merged**